### PR TITLE
Basic SAM2 AutomaticMaskGeneration example server

### PR DESCRIPTION
**Pull request by cpuhrsch**
Run
```
$ python server.py
INFO:root:Loading model: checkpoints/sam2_hiera_large.pt
INFO:root:Loaded checkpoint sucessfully
INFO:root:Running warmup.
INFO:root:For numpy array image, we assume (HxWxC) format
INFO:root:Computing image embeddings for the provided image...
INFO:root:Image embeddings computed.
[...]
INFO:root:For numpy array image, we assume (HxWxC) format
INFO:root:Computing image embeddings for the provided image...
INFO:root:Image embeddings computed.
INFO:root:For numpy array image, we assume (HxWxC) format
INFO:root:Computing image embeddings for the provided image...
INFO:root:Image embeddings computed.
INFO:root:Warmup took 6.490015029907227s.
INFO:     Started server process [3042910]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:5000 (Press CTRL+C to quit)
```
and then open example.html to get
Screenshot 2024-10-08 at 1 29 52 PM

which results in
```
INFO:     Uvicorn running on http://127.0.0.1:5000 (Press CTRL+C to quit)
INFO:root:For numpy array image, we assume (HxWxC) format
INFO:root:Computing image embeddings for the provided image...
INFO:root:Image embeddings computed.
Took 1.9002370834350586 to generate a mask for input image.
INFO:     127.0.0.1:51100 - "POST /upload HTTP/1.1" 200 OK
```

on 1 A100 GPU.

NOTE: This is basic starter code. Follow up PRs will add more docs and performance improvements.

### CODE REVIEW

**Review by vkuzo — APPROVED** [review]
stamp!

File: torchao/prototype/server/amg_example/README.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+To run this example you need to download the vit_h checkpoint and put it into a local folder named checkpoints
```
Comment (msaroufim) [inline torchao/prototype/server/amg_example/README.md:1]: nit: but prototype is sort of losing its meaning lol. This is closer to an example so mind creating a new examples folder?

File: torchao/prototype/server/amg_example/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+
+You can find the checkpoint for vit_h here: https://dl.fbaipublicfiles.com/segment_anything/sam_vit_h_4b8939.pth
```
Comment (msaroufim) [inline torchao/prototype/server/amg_example/README.md:3]: nit: wget

File: torchao/prototype/server/amg_example/amg_example.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+torch._dynamo.config.capture_dynamic_output_shape_ops = True
+mask_generator._process_batch = torch.compile(
+    mask_generator._process_batch,
+    mode="max-autotune-no-cudagraphs",
+    fullgraph=True,
+    dynamic=True,
+)
+
+# with torch.backends.cuda.sdp_kernel(enable_cudnn=False): #, enable_math=False, enable_mem_efficient=False):
+with torch.backends.cuda.sdp_kernel(enable_cudnn=True): #, enable_math=False, enable_mem_efficient=False):
```
Comment (msaroufim) [inline torchao/prototype/server/amg_example/amg_example.py:96]: H100?

File: torchao/prototype/server/amg_example/amg_example.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    
+    # Save a GPU trace
+    profiler_runner(f"amg_example_trace.json.gz", mask_generator.generate, image)
+    
+    # Write out memory usage
+    max_memory_allocated_bytes = torch.cuda.max_memory_allocated()
+    _, total_memory = torch.cuda.mem_get_info()
+    max_memory_allocated_percentage = int(100 * (max_memory_allocated_bytes / total_memory))
+    max_memory_allocated_bytes = max_memory_allocated_bytes >> 20
+    print(f"memory(MiB): {max_memory_allocated_bytes} memory(%): {max_memory_allocated_percentage}")
```
Comment (msaroufim) [inline torchao/prototype/server/amg_example/amg_example.py:133]: can we avoid checking in binary data? Mind hosting the images on imgur or github instead, makes cloning the repo slower

File: torchao/prototype/server/amg_example/server.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+from torch._inductor import config as inductorconfig
+inductorconfig.triton.unique_kernel_names = True
+inductorconfig.coordinate_descent_tuning = True
+inductorconfig.coordinate_descent_check_all_directions = True
+
+# torch.set_float32_matmul_precision('high')
```
Comment (msaroufim) [inline torchao/prototype/server/amg_example/server.py:31]: how come?

File: torchao/prototype/server/amg_example/example.html [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            display: none;
+        }
+    </style>
+</head>
+<body>
+    <h1>Upload an Image and Display the Response</h1>
+    <form id="uploadForm">
+        <label for="image">Choose an image to upload:</label><br>
+        <input type="file" id="image" name="image" accept="image/*" required><br><br>
+        <input type="submit" value="Upload Image">
```
Comment (msaroufim) [inline torchao/prototype/server/amg_example/example.html:21]: is there a way to skip need to click the upload image button? basically make upload automatic if people drag the image to the browser window

**Review by msaroufim — COMMENTED** [review]
some nits and minor suggestions

File: torchao/prototype/server/amg_example/README.md (hunk 1 above)
Comment (cpuhrsch) [reply torchao/prototype/server/amg_example/README.md:1]: I'll refactor it to be closer to that in the next PR if that works.

File: torchao/prototype/server/amg_example/amg_example.py (hunk 3 above)
Comment (cpuhrsch) [reply torchao/prototype/server/amg_example/amg_example.py:96]: No, this also was an issue on A100

File: torchao/prototype/server/amg_example/amg_example.py (hunk 4 above)
Comment (cpuhrsch) [reply torchao/prototype/server/amg_example/amg_example.py:133]: Yes, will do.

File: torchao/prototype/server/amg_example/example.html (hunk 6 above)
Comment (cpuhrsch) [reply torchao/prototype/server/amg_example/example.html:21]: Probably, good idea for the next iteration.

File: torchao/prototype/server/amg_example/README.md (hunk 1 above)
Comment (cpuhrsch) [reply torchao/prototype/server/amg_example/README.md:1]: I'll just move it into examples right now.

**Pull request merged**